hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java

```diff
@@ -1633,6 +1636,53 @@ public DatanodeInfo[] getDataNodeStats(final DatanodeReportType type)
    * @see org.apache.hadoop.hdfs.protocol.ClientProtocol#setSafeMode(
    *    HdfsConstants.SafeModeAction,boolean)
    */
+  @Override
+  public boolean setSafeMode(SafeModeAction action)
+    throws IOException {
+    return setSafeMode(action, false);
+  }
+
+  /**
+   * Enter, leave or get safe mode.
+   *
+   * @param action
+   *          One of SafeModeAction.ENTER, SafeModeAction.LEAVE and
+   *          SafeModeAction.GET
+   * @param isChecked
+   *          If true check only for Active NNs status, else check first NN's
+   *          status
+   */
+  @Override
+  public boolean setSafeMode(SafeModeAction action, boolean isChecked)
+    throws IOException {
+    return dfs.setSafeMode(convertToClientProtocolSafeModeAction(action), isChecked);
```

Comment:
how about translating to the hdfs action and calling it on *this* class. This ensures that existing subclasses, especially ViewDistributedFileSystem get the request forwarded. Without it, `ViewDistributedFileSystem` is broken

Reply:
to confirm your suggestion, should I just the replace `dfs` to `this` like below?

```suggestion
    return this.setSafeMode(convertToClientProtocolSafeModeAction(action), isChecked);
```

Reply:
yes, as that way ViewDistributedFileSystem will be invoked as before and forward it using its rules